### ISSUE TITLE
Non-deterministic size of an empty window

### ISSUE THREAD

**Issue by matklad** [sixtyfpsui/sixtyfps#235]
Commit: https://github.com/matklad/120wpm/commit/09aa9dd01257ca6634b6e28c62465dfe104d8e5f
Command: `cargo run --release`

This commit should create an empty window 1000x500 pixels in size. However, the size is non-deterministic for some reason:

![image](https://user-images.githubusercontent.com/1711539/119106450-0aa4de80-ba27-11eb-98c4-ab543f6daf47.png)

**Comment by matklad** [comment]
Minimized in https://github.com/matklad/120wpm/commit/32876a9c8e2e33a7d7d41d1f020fe647f66fb994 using 0.0.6 version

![image](https://user-images.githubusercontent.com/1711539/119107120-a20a3180-ba27-11eb-9ed3-7b909994b749.png)

**Comment by matklad** [comment]
`viewer ui/main.60 & viewer ui/main.60 ` also reproduces this. BTW while we are at it:

* `viewer` binary name is a bit unintuitive, it took me a second to realize why `sixtyfps-viewer` command doesn't work in the shell
* publish `viewer` to crates.io :heart_eyes:

**Comment by tronical** [comment]
Interesting. I can't reproduce this with 0.0.6 or HEAD, using Ubuntu 21.04 with a GNOME session as well as KDE, with and without `SIXTYFPS_BACKEND=GL` (using `cargo run --bin viewer -- ~/src/120wpm/ui/main.60`)

It's good that you can reproduce this straight using the viewer. (your vote on publishing to crates.io noted and seconded :-)

In a effort to circle in on the culprit, does it make any difference for you if you set `width` and `height` in your `MainWindow` to `1000px`/`500px` directly instead of through the variables? Is this reproducible for you with a plain
```
export App := MainWindow {
    width: 800px;
    height: 600px;
}
```
?

**Comment by ogoffart** [comment]
The way windows are sized actually changed a lot recently.
In master, setting width or height on the window should set a fixed sized.
The initial size can be set with `preferred-width` / `preferred-height` 

But i also couldn't reproduce this problem in master.
Is it the Qt backend or the GL backend? Is that X11 or wayland?  What sha1 of sixtyfps are you using?

**Comment by matklad** [comment]
Yep, 

```
App := Window {
  height: 1000px;
  width: 500px;
}
```

and viewer (https://github.com/sixtyfpsui/sixtyfps/#1aed36d1)) is enough to repro the issue. I am using OpenGL backend on nixos on X11, so there might be all kinds of weird dynamic library shenanigans. For the same nixos reason, I don't think I can try Qt backend easily :( I've tried running under `strace`, and that made the bug go away :(

Is there some kind of env var/build flag I can set to get debug logs?

**Comment by ogoffart** [comment]
I can reproduce now

**Comment by matklad** [comment]
Great! Just for fun, I've added some debug prints:

```
diff --git a/sixtyfps_runtime/rendering_backends/gl/lib.rs b/sixtyfps_runtime/rendering_backends/gl/lib.rs
index e86802a8..87107d69 100644
--- a/sixtyfps_runtime/rendering_backends/gl/lib.rs
+++ b/sixtyfps_runtime/rendering_backends/gl/lib.rs
@@ -423,6 +423,7 @@ fn new_renderer(
         default_font_properties: &Pin<Rc<Property<FontRequest>>>,
     ) -> GLItemRenderer {
         let size = self.window().inner_size();
+        eprintln!("size = {:?}", size);
 
         #[cfg(not(target_arch = "wasm32"))]
         {

```

Output:

```
12:43:29|~/tmp
λ viewer main.60
size = PhysicalSize { width: 800, height: 600 }
size = PhysicalSize { width: 800, height: 600 }

12:43:32|~/tmp
λ viewer main.60
size = PhysicalSize { width: 750, height: 1500 }
```

**Comment by ogoffart** [comment]
Ok, I debugged the problem.

The root cause is a problem in the winit API, we are supposed to set the size on the window before having a window instance.
But the computation of the bindings themselves need a window instance for the text size computation or the scale factor property.
Anyway, what we do is to first create a window, then set the initial size.
But winit will always send the size to the WM to 800x600 if we did not specify the size in the WindowBuilder:

https://github.com/rust-windowing/winit/blob/b371b406d5fea019c074a3b6b68a765e3f2d9c99/src/platform_impl/linux/x11/window.rs#L156-L159

So here is basicaly what happens:

 1. The window is shown, winit sends 800x600 size to the X11 server
 2. Se sets the width / height property base on the bindings, (in `apply_window_properties`) this sends the new size to the X11 server
 3. We recieve a ResizeEvent from winit with 800x600,   so we update the height/width property accordingly to track the actual size (as if it was resized by the user)
 4. We weceive aother ResizeEvent with the size set at step 2. we update the height/width property accordingly
 
This is what happens most of the time ans also why the window is flickering at the begining.

However, for some reason, sometimes, step 3 and 4 happens in the same eventloop iteration, sometimes it happens in two different iterations.

I forgot to mention the step where we set the window constraint, and this is set to the current value of the width/height property.
This ends up being set at the next iteration of the event loop, after the width or height are changed. As a result, depending if the x11 server sends the two resize event in the same iteration of the eventloop or not, we end up setting the size constraints differently.

First thing to fix would be to set the constraints immediately and not on the next iteration of the event loop, that should hopefully fix the symptoms.

I don't know if it would make sense to fix winit not to send a hardcoded size and let the WM do its job
